File: mmKyber-c/mm_sample.c
//  mm_sample.c
//  === Simple Samplers (XXX: the Gaussian ones are placeholders.)

#include <math.h>
#include "mm_param.h"
#include "mm_sample.h"

//  Uniform sampler [0, q-1]

#define MM_Q_SZ     ((MM_LOGQ + 7) / 8)
#define MM_Q_MASK   ((1 << MM_LOGQ) - 1)

void poly_unif(int32_t *r, sha3_t *kec)
{
    uint8_t h[4] = { 0 };
    int i;
    int32_t x;

    i = 0;
    while (i < MM_D) {
        do {
            sha3_squeeze(kec, h, MM_Q_SZ);
            x = get32u_le(h) & MM_Q_MASK;
        } while (x >= MM_Q);
        r[i++] = x;
    }
}

//  Uniform sampler [-nu, nu]
/*
#define MM_NU_S     (2 * (MM_NU) + 1)
#define MM_NU_MASK  ((1 << MM_NU_BITS) - 1)

void poly_nu(int32_t *r, sha3_t *kec)
{
    uint8_t h[8] = { 0 };
    int i, l;
    int32_t x;
    uint64_t t;

    i = 0;
    l = 0;
    while (i < MM_D) {
        if (l < MM_NU_BITS) {
            sha3_squeeze(kec, h, 8);
            t = get64u_le(h);
            l = 64;
        }
        x = t & MM_NU_MASK;
        t >>= MM_NU_BITS;
        l -= MM_NU_BITS;
        if (x < MM_NU_S) {
            r[i++] = x - MM_NU;
        }
    }
}
*/

#if (MM_NU_BAR == 2)

//  Sample a polynomial in binary uniform set {0,1}.

void sample_nu(uint8_t *r, sha3_t *kec)
{
    sha3_squeeze(kec, r, MM_D / 8);
}

//  Decode a binary polynomial from bytes..

void poly_nu(int32_t *r, const uint8_t *s)
{
    int i, j, x;

    for (i = 0; i < MM_D; i += 8) {
        x = *s++;
        for (j = 0; j < 8; j++) {
            r[j] = (x >> j) & 1;
        }
        r += 8;
    }
}

#elif (MM_NU_BAR == 3)

//  Sample a polynomial in ternary uniform set {-1,0,1}.

void sample_nu(uint8_t *r, sha3_t *kec)
{
    int i;

    i = 0;
    while (i < (MM_D + 4) / 5) {
        sha3_squeeze(kec, &r[i], 1);
        if (r[i] < 243) {
            i++;
        }
    }
}

//  Decode a ternary polynomial from 52 bytes in [0, 243).

void poly_nu(int32_t *r, const uint8_t *s)
{
    int i;
    uint32_t x, a, b;

    for (i = 0; i < 51; i++) {
        x = s[i];

        a = x * 0xAAAB;
        b = (a >> 13) & 3;
        r[0] = (int32_t) b - 1;
        x = (x - b) * 0xAAAB;

        a = x * 0xAAAB;
        b = (a >> 13) & 3;
        r[1] = (int32_t) b - 1;
        x = (x - b) * 0xAAAB;

        a = x * 0xAAAB;
        b = (a >> 13) & 3;
        r[2] = (int32_t) b - 1;
        x = (x - b) * 0xAAAB;

        a = x * 0xAAAB;
        b = (a >> 13) & 3;
        r[3] = (int32_t) b - 1;
        x = (x - b) * 0xAAAB;

        a = x * 0xAAAB;
        b = (a >> 13) & 3;
        r[4] = (int32_t) b - 1;

        r += 5;
    }
    x = s[51];
    a = x * 0xAAAB;
    b = (a >> 13) & 3;
    r[0] = (int32_t) b - 1;
}

#endif

/*
=== IMPORANT NOTICE: This sampler is approximate and not constant-time;
    it is a placeholder implementation. A more appropriate Discrete Gaussian
    sampler is required in practice.
*/

//  Gaussian sampler

void poly_gauss(int32_t *r, sha3_t *kec, double sigma)
{
    int i;
    double cs2, d63;
    double x, y, w;
    uint8_t h[16] = { 0 };

    cs2 = 1.0 / 6.0 - 2.0 * (sigma * sigma);
    d63 = ldexp(1.0, -63);

    i = 0;
    while (i < MM_D) {
        sha3_squeeze(kec, h, 16);
        x = d63 * ((double) get64u_le(h)) - 1.0;
        y = d63 * ((double) get64u_le(h + 8)) - 1.0;
        w = x*x + y*y;
        if (w > 0.0 && w <= 1.0) {
            w = sqrt( cs2 * log(w) / w );
            r[i++] = rint( x * w );
            r[i++] = rint( y * w );
        }
    }
}

File: mmKyber-py/mm_sample.py
#   mm_sample.py
#   === Uniform and Discrete Gaussian samplers

import math
from mm_ring import MM_D, MM_Q

def poly_unif(xof, q=MM_Q, d=MM_D):
    """ Sample an uniform integer polynomial in [0,q-1]. """
    bits_q  = int(q - 1).bit_length()
    byte_q  = (bits_q + 7) // 8
    mask_q  = (1 << bits_q) - 1
    r   = []
    while len(r) < d:
        x   = int.from_bytes(xof.read(byte_q), byteorder='little') & mask_q
        if x < q:
            r += [x]
    return r

def sample_tri(xof, d=MM_D):
    """ Sample a polynomial in ternary uniform set {-1,0,1}."""
    b = b''
    i = 0
    while i < d:
        x = xof.read(1)[0]
        if x < 243:
            b += bytes([x])
            i += 5
    return b

def sample_bin(xof, d=MM_D):
    """ Sample a polynomial in the binary set {0,1}."""
    return xof.read((d + 7) // 8);

def poly_tri(s, d=MM_D):
    """ Decode a ternary polynomial from bytes in [0,243). """
    v = []
    for x in s:
        for i in range(5):
            v += [ (x % 3) - 1 ]
            x //= 3
    return v[:d]

def poly_bin(s, d=MM_D):
    """ Decode a binary polynomial from bytes. """
    v = []
    for x in s:
        v += [ (x >> i) & 1  for  i in range(8) ]
    return v[:d]


"""
=== IMPORANT NOTICE: This sampler is approximate and not constant-time;
    it is a placeholder implementation. A more appropriate Discrete Gaussian
    sampler is required in practice.

Firstly we note that "Gaussian width" is related to standard distribution
by sigma = gw / sqrt(2*Pi).

Rounded Gaussian sampler uses narrower sigma' = sqrt(sigma^2 - 1/12)
in the continuous sampler to get the desired distribution sigma.
For intuition, note that 1/12 is the variance of [-0.5, +0.5]
uniform distribution; models the rounding effect when sigma is larger.
Sect A.3 in https://eprint.iacr.org/2017/1025.pdf
Uses S. Janson 2006: http://dx.doi.org/10.1214/009117906000000232
For a Renyi argument, see Appendix E in https://eprint.iacr.org/2024/184.pdf

    orig    cs2 =   -2 * sigma'^2
    substitute      -2 * (sigma^2 - 1/12)
    simplify        1/6 - 2*sigma^2
    simplify        1/6 - 2*(gw/sqrt(2*Pi))^2
                    1/6 - gw^2/Pi

"""

def poly_gauss(xof, gw, d=MM_D):
    """ Discrete gaussian with width gw = sigma/sqrt(2*Pi). """
    cs2 =   1.0/6.0 - (gw*gw)/math.pi       #   adjust sigma
    d63 =   0.5**63                         #   scaling
    r   =   []
    while len(r) < d:
        x   = d63 * int.from_bytes(xof.read(8), byteorder='little') - 1.0
        y   = d63 * int.from_bytes(xof.read(8), byteorder='little') - 1.0
        w   = x*x + y*y
        if w > 0.0 and w <= 1.0:
            w   = math.sqrt( cs2 * math.log(w) / w )
            r   += [ round(x * w) ]
            r   += [ round(y * w) ]
    return r

